### DISCUSSION TITLE
How's memory usage and startup time with the newer versions, especially on ARM(64)?

### DISCUSSION THREAD

**Discussion by jonaswinkler** [jonaswinkler/paperless-ng#554] [General]
I've adjusted a couple things here and there in the application up until 1.1.3 that should decrease overal RAM usage. Are there any noticeable differences, especially on low powered devices?

* When consuming documents, paperless now frees all memory of the consumption process after it's done.
* I've adjusted various code segemnts so that they only load modules when absolutely necessary. For example, paperless requires a certain parser component for performing OCR on PDF documents and getting metadata from PDF documents (that's the metadata tab on the web site). Before the change, paperless would load the entire `ocrmypdf` module on the webserver process, even though it only required certain parts of that for metadata extraction.

Furthermore, I've gained some more insights about memory usage:

* Redis adds about +4MB over here.
* Tika+Gotenberg require about +150MB when idling, and about +500MB when working on a document.
* When automatic matching is used, paperless needs to load a lot of additional dependencies and data to to perform the matching. Aprox +100MB to +150MB, depending on the size of the document collection. When automatic matching is not used and has never been used in the past, absolutely no additional impact.
* I've tested various ASGI web servers (`daphne`, `uvicorn` standalone, `uvicorn` as a worker process for `gunicorn`), and when `PAPERLESS_WEBSERVER_WORKERS` is set to 1, the current gunicorn+uvicorn setup is pretty much on par with the other options.
* With `PAPERLESS_WEBSERVER_WORKERS` > 1 (2 is default), each worker process will load the application separately. These also won't load the entire application immediately, but rather load the parts that are required to serve specific requests, so memory usage is expected to slowly creep up until everything is loaded. Configuring this to a higher value is faster, but more memory will be used.

**Comment by writepatrick** [comment]
Hi Jonas,
I am wondering why nobody wrote their experiences with low powered devices like a RaspberryPi. 

My setup:
- Raspberry Pi 4 4GB
- Raspberry Pi OS Lite
- passive cooling
- official USB-C power-supply
- external 2,5" Hard-Drive via USB3,0 (no additional power)

I am currently running paperless-ng 1.3.0 with docker-compose and using postgres database without Tika.
I disabled inotify since the network share didn't like it (netatalk AFPS) and paperless-ng tried to process new documents before they were completed.
Everything else is default configuration.

When idling the RPi is using about 400-500MB  and the overall performance is pretty good. I am ingesting most files via GeniusScan sometimes with OCR on the Phone (which was recommended here #254 ) and transfering via FTP. The OCR processing on the RPi 4 is quite good and the results are superb (better as the apps OCR)
The automatic matching works astonishingly good after the sample size was sufficient.

My wife and I are really happy to have found a solution for our paper-problem. (and I have an excuse to fiddle around with the RPi)

I am really glad you took paperless to a new level! Keep it up.

**Reply by jonaswinkler** [reply]
Thanks for the detailed report! I've only got a RPi 3 over here which I don't use for paperless at all, so I don't have any idea how well it runs. I'm glad it works out for you this well. :)
